platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala

```diff
@@ -1,57 +1,96 @@
+        (x.edata.eks.asInstanceOf[Map[String, AnyRef]].get("total_ts").get.asInstanceOf[Double])
+      }.sum
+      val total_ts = if (prevProfileData.total_ts.isEmpty) current_ts else current_ts + prevProfileData.total_ts.get
+      val current_launches = events.currentData.map { x =>
+        (x.edata.eks.asInstanceOf[Map[String, AnyRef]].get("total_launches").get.asInstanceOf[Number].longValue())
+      }.sum
+      val total_launches = if (prevProfileData.total_launches.isEmpty) current_launches else current_launches + prevProfileData.total_launches.get
+      val avg_ts = if (total_launches == 0) total_ts else CommonUtil.roundDouble(total_ts / total_launches, 2)
+      DeviceProfileOutput(events.index.device_id, Option(first_access), Option(last_access), Option(total_ts), Option(total_launches), Option(avg_ts), prevProfileData.device_spec, prevProfileData.uaspec, prevProfileData.state, prevProfileData.city, prevProfileData.country, prevProfileData.country_code, prevProfileData.state_code, prevProfileData.state_custom, prevProfileData.state_code_custom, prevProfileData.district_custom, prevProfileData.fcm_token, prevProfileData.producer_id, prevProfileData.user_declared_state, prevProfileData.user_declared_district, prevProfileData.api_last_updated_on)
     }
+  }
+
+  override def postProcess(data: RDD[DeviceProfileOutput], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[Empty] = {
+    try{
+    data.foreachPartition {partitions: Iterator[DeviceProfileOutput] =>
+      partitions.foreach{ f =>
+        println("******************Inside post process foreach*********************")
+        val keyList: List[String] = List("first_access", "last_access", "updated_date", "api_last_updated_on")
+        val first_access = f.first_access.get
+        val last_access = f.last_access.get
+        val updated_date = f.updated_date.get
+        val api_last_updated = f.api_last_updated_on.getOrElse(new Timestamp(System.currentTimeMillis()))
+        val fieldMap = JSONUtils.deserialize[Map[String, Any]](JSONUtils.serialize(f))
+        val filteredMap = fieldMap.--(keyList)
+        val resultMap = filteredMap ++ Map("first_access" -> first_access, "last_access" -> last_access, "updated_date" -> updated_date, "api_last_updated_on" -> api_last_updated)
 
-    override def postProcess(data: RDD[DeviceProfileOutput], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[Empty] = {
-        data.saveToCassandra(Constants.DEVICE_KEY_SPACE_NAME, Constants.DEVICE_PROFILE_TABLE);
-        sc.makeRDD(List(Empty()));
+        val columns = resultMap.keySet.mkString(",")
+        val values = resultMap.values.mkString("','")
+        val queryReq = s"""INSERT INTO $table ($columns) VALUES ('$values') ON CONFLICT (device_id) DO UPDATE SET ($columns) = ('$values')"""
+        postgresDB.insertDataToPostgresDB(queryReq)
+      }
+    }
+    }
+    finally {
+      postgresDB.closeConnection
```

Comment:
This is not right - opening and closing a connection should be in each partition